src/uct/ugni/base/ugni_iface.c

```diff
@@ -0,0 +1,299 @@
+
+void uct_ugni_base_desc_init(void *mp_context, void *obj, void *chunk, void *arg)
+{
+  uct_ugni_base_desc_t *base = (uct_ugni_base_desc_t *) obj;
+  /* zero base descriptor */
+  memset(base, 0 , sizeof(*base));
+}
+
+void uct_ugni_base_desc_key_init(uct_iface_h iface, void *obj, uct_mem_h memh)
+{
+  uct_ugni_base_desc_t *base = (uct_ugni_base_desc_t *)obj;
+  /* call base initialization */
+  uct_ugni_base_desc_init(iface, obj, NULL, NULL);
+  /* set local keys */
+  base->desc.local_mem_hndl = *(gni_mem_handle_t *)memh;
+}
+
+void uct_ugni_progress(void *arg)
+{
+    gni_cq_entry_t  event_data = 0;
+    gni_post_descriptor_t *event_post_desc_ptr;
+    uct_ugni_base_desc_t *desc;
+    uct_ugni_iface_t * iface = (uct_ugni_iface_t *)arg;
+    gni_return_t ugni_rc;
+
+    ugni_rc = GNI_CqGetEvent(iface->local_cq, &event_data);
+    if (GNI_RC_NOT_DONE == ugni_rc) {
+        return;
+    }
+
+    if ((GNI_RC_SUCCESS != ugni_rc && !event_data) || GNI_CQ_OVERRUN(event_data)) {
+        ucs_error("GNI_CqGetEvent falied. Error status %s %d ",
+                  gni_err_str[ugni_rc], ugni_rc);
+        return;
+    }
+
+    ugni_rc = GNI_GetCompleted(iface->local_cq, event_data, &event_post_desc_ptr);
+    if (GNI_RC_SUCCESS != ugni_rc && GNI_RC_TRANSACTION_ERROR != ugni_rc) {
+        ucs_error("GNI_GetCompleted falied. Error status %s %d %d",
+                  gni_err_str[ugni_rc], ugni_rc, GNI_RC_TRANSACTION_ERROR);
+        return;
+    }
+
+    desc = (uct_ugni_base_desc_t *)event_post_desc_ptr;
+    ucs_trace_async("Completion received on %p", desc);
+
+    if (NULL != desc->comp_cb) {
+        uct_invoke_completion(desc->comp_cb);
+    }
+    --iface->outstanding;
+    --desc->ep->outstanding;
+
+    if (ucs_likely(desc->not_ready_to_free == 0)) {
+        ucs_mpool_put(desc);
+    }
+    return;
+}
+
+ucs_status_t uct_ugni_iface_flush(uct_iface_h tl_iface)
+{
+    uct_ugni_iface_t *iface = ucs_derived_of(tl_iface, uct_ugni_iface_t);
+    if (0 == iface->outstanding) {
+        return UCS_OK;
+    }
+    uct_ugni_progress(iface);
+    return UCS_ERR_NO_RESOURCE;
+}
+
+ucs_status_t uct_ugni_iface_get_address(uct_iface_h tl_iface,
+                                               struct sockaddr *addr)
+{
+    uct_ugni_iface_t *iface = ucs_derived_of(tl_iface, uct_ugni_iface_t);
+    uct_sockaddr_ugni_t *iface_addr = (uct_sockaddr_ugni_t*)addr;
+
+    iface_addr->sgni_family = UCT_AF_UGNI;
+    iface_addr->nic_addr    = iface->nic_addr;
+    iface_addr->domain_id   = iface->domain_id;
+    return UCS_OK;
+}
+
+int uct_ugni_iface_is_reachable(uct_iface_h tl_iface, const struct sockaddr *addr)
+{
+    const uct_sockaddr_ugni_t *iface_addr = (const uct_sockaddr_ugni_t*)addr;
+
+    if (iface_addr->sgni_family != UCT_AF_UGNI) {
+        return 0;
+    }
+
+    return 1;
+}
+
+static ucs_status_t get_cookie(uint32_t *cookie)
+{
+    char           *cookie_str;
+    char           *cookie_token;
+
+    cookie_str = getenv("PMI_GNI_COOKIE");
```

Comment:
getenv/setenv is not thread-safe API.
I think we have ucx_getenv/setenv or similar which should be used.